File: 백준/Silver/1182. 부분수열의 합/부분수열의 합.py
import sys
from itertools import combinations
input = sys.stdin.readline

n, s = map(int, input().split())
nums = list(map(int,input().split()))
answer = 0

for i in range(1, n+1):
    for sub in combinations(nums, i):
        if sum(sub) == s:
            answer += 1

print(answer)

File: 백준/Silver/1182. 부분수열의 합/부분수열의 합.js
const input = require('fs').readFileSync(process.platform === 'linux' ? '/dev/stdin' : __dirname + '/input.txt').toString().split('\n')

const [n, s] = input[0].split(' ').map(Number)
const nums = input[1].split(' ').map(Number)
let answer = 0

for (let i = 1; i < n + 1; i++) {
  combination(nums, i, 0, [])
}
console.log(answer)

function combination(arr, r, index, cur) {
  if (cur.length === r) {
    const sum = cur.reduce((a, b) => a + b)
    if (sum === s) answer++
    return
  }
  
  for (let i = index; i < arr.length; i++) {
    cur.push(arr[i])
    combination(arr, r, i + 1, cur)
    cur.pop()
  }
}

